File: jmlparser-jml-tools/src/main/java/com/github/jmlparser/smt/ArithmeticTranslator.java
package com.github.jmlparser.smt;

import com.github.javaparser.ast.NodeList;
import com.github.javaparser.ast.body.Parameter;
import com.github.javaparser.ast.expr.*;
import com.github.javaparser.resolution.types.ResolvedType;
import com.github.jmlparser.smt.model.SExpr;
import com.github.jmlparser.smt.model.SmtType;

import java.math.BigInteger;
import java.util.List;
import java.util.stream.Collectors;

/**
 * @author Alexander Weigl
 * @version 1 (01.07.22)
 */
public interface ArithmeticTranslator {
    SExpr binary(BinaryExpr.Operator operator, SExpr left, SExpr right);

    SExpr makeChar(CharLiteralExpr n);

    SExpr unary(UnaryExpr.Operator operator, SExpr accept);

    SExpr makeLong(LongLiteralExpr n);

    SExpr makeInt(IntegerLiteralExpr n);

    SExpr makeInt(BigInteger i);

    SExpr makeIntVar();

    default List<? extends SExpr> getVariable(NodeList<Parameter> variables) {
        return variables.stream().map(this::getVariable).collect(Collectors.toList());
    }

    SExpr getVariable(Parameter jmlBoundVariable);

    SExpr makeBoolean(boolean value);

    SmtType getType(ResolvedType asPrimitive);

    SExpr arrayLength(SExpr obj);

    SExpr makeInt(long i);

    SExpr makeVar(ResolvedType rtype);
}

File: tools/smt/src/main/kotlin/io/github/jmltoolkit/smt/ArithmeticTranslator.kt
package io.github.jmltoolkit.smt

import com.github.javaparser.ast.NodeList
import com.github.javaparser.ast.body.Parameter
import com.github.javaparser.ast.expr.*
import com.github.javaparser.resolution.types.ResolvedType
import io.github.jmltoolkit.smt.model.SExpr
import io.github.jmltoolkit.smt.model.SmtType
import java.math.BigInteger

/**
 * @author Alexander Weigl
 * @version 1 (01.07.22)
 */
interface ArithmeticTranslator {
    fun binary(operator: BinaryExpr.Operator, left: SExpr, right: SExpr): SExpr

    fun makeChar(n: CharLiteralExpr): SExpr

    fun unary(operator: UnaryExpr.Operator, accept: SExpr): SExpr

    fun makeLong(n: LongLiteralExpr): SExpr

    fun makeInt(n: IntegerLiteralExpr): SExpr

    fun makeInt(i: BigInteger): SExpr

    fun makeIntVar(): SExpr

    fun getVariable(variables: NodeList<Parameter>): List<SExpr> {
        return variables.map { getVariable(it) }
    }

    fun getVariable(jmlBoundVariable: Parameter): SExpr

    fun makeBoolean(value: Boolean): SExpr

    fun getType(asPrimitive: ResolvedType): SmtType

    fun arrayLength(obj: SExpr): SExpr

    fun makeInt(i: Long): SExpr

    fun makeVar(rtype: ResolvedType): SExpr
}
